core/drivers/crypto/se050/adaptors/utils/utils.c

```diff
@@ -0,0 +1,173 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) Foundries Ltd. 2020 - All Rights Reserved
+ * Author: Jorge Ramirez <jorge@foundries.io>
+ */
+
+#include <crypto/crypto.h>
+#include <se050.h>
+#include <string.h>
+#include <util.h>
+
+#define OID_MIN			((uint32_t)(0x00000001))
+#define OID_MAX			((uint32_t)(OID_MIN + 0x7BFFFFFE))
+#define NBR_OID			((uint32_t)(OID_MAX - OID_MIN))
+#define IS_WATERMARKED(x)	(((x) & WATERMARKED(0)) == WATERMARKED(0))
+
+static void delete_transient_objects(void)
+{
+	Se05xSession_t *ctx = NULL;
+	uint8_t more = kSE05x_MoreIndicator_NA;
+	uint8_t *list = NULL;
+	size_t len = 1024;
+	smStatus_t status  = SM_NOT_OK;
+	uint16_t offset = 0;
+	uint32_t id = 0;
+	size_t i = 0;
+
+	if (!se050_session)
+		return;
+
+	ctx = &se050_session->s_ctx;
+
+	list = calloc(1, len);
+	if (!list)
+		return;
+	do {
+		status = Se05x_API_ReadIDList(ctx, offset, 0xFF, &more,
+					      list, &len);
+		if (status != SM_OK) {
+			free(list);
+			return;
+		}
+
+		offset = len;
+		for (i = 0; i < len; i += 4) {
+			id = (list[i + 0] << (3 * 8)) |
+			     (list[i + 1] << (2 * 8)) |
+			     (list[i + 2] << (1 * 8)) |
+			     (list[i + 3] << (0 * 8));
+
+			if (id >= OID_MAX || id == 0)
+				continue;
+
+			if (id & BIT(0))
+				Se05x_API_DeleteSecureObject(ctx, id);
+		}
+	} while (more == kSE05x_MoreIndicator_MORE);
+
+	free(list);
+}
+
+static uint32_t generate_oid(sss_key_object_mode_t mode)
+{
+	uint32_t oid = OID_MIN;
+	uint32_t random = 0;
+	size_t i = 0;
+
+	for (i = 0; i < NBR_OID; i++) {
```

Comment:
in theory it shouldn't need to a single loop  but it could be that the OID is in use already (ie, the id could have been created in a previous session months ago for instance) ...hence the loop